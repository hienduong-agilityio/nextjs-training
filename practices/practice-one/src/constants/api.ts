export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

const ENV = {
  ROOT_API: process.env.NEXT_PUBLIC_SERVER_URL,
  USER_ID: process.env.NEXT_PUBLIC_USER_ID,
};

export const API_URL = {
  PRODUCT: `${ENV.ROOT_API}/product`,
  CART: `${ENV.ROOT_API}/carts`,
};

export const USER_ID = ENV.USER_ID;
