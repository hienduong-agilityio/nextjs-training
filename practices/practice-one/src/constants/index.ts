export * from './api';
export * from './category';
export * from './searchParam';
export * from './router';
