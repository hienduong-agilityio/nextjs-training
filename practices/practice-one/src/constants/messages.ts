// Toast Messages
export const TOAST_MESSAGES = {
  ADD_SUCCESS: 'Product added to cart successfully!',
  ADD_MAX_QUANTITY: 'Unable to add product. Maximum quantity exceeded!',
  ADD_FAILED: 'Failed to add product to cart. Please try again!',
  DELETE_SUCCESS: 'Product removed from cart successfully!',
  DELETE_FAILED: 'Failed to remove product from cart. Please try again!',
  API_ERROR: 'API Error. Please try again!',
};

export const TOAST_TYPES = {
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
};
