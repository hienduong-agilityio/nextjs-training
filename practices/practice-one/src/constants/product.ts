export const ALL_CATEGORIES = {
  ALL: 'all',
  BEAUTY: 'beauty',
  FRAGRANCES: 'fragrances',
  FURNITURE: 'furniture',
  GROCERIES: 'groceries',
};

export const SORT_PRODUCT_OPTIONS = ['name', 'price', 'rating'];
