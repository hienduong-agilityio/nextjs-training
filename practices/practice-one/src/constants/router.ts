export const ROUTE = {
  ROOT: '/',
  PRODUCTS: '/products',
  PRODUCT: '/product',
  CART: '/cart',
  CHECKOUT: '/checkout',
  COLLECTION: '/collection',
  ABOUT_US: '/about-us',
  INFORMATION: '/information',
  PRIVACY_POLICY: '/privacy-policy',
  TERM_AND_CONDITIONS: '/terms-and-conditions',
};
