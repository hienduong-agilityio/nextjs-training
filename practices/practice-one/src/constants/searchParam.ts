export const SEARCH_PARAMS = {
  ACTIVE_TAB: 'activeTab',
  LIMIT: 'limit',
  PAGE: 'page',
  SORT_BY: 'sortBy',
  CATEGORY: 'category',
  SEARCH: 'search',
} as const;
