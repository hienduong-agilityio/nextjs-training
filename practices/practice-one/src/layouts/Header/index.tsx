import { Suspense } from 'react';

// Icon
import { CartIcon, LogoIcon, ProfileIcon } from '@/icons';

// Components
import { SearchBox, SearchModal } from '@/components';
import Link from 'next/link';

// Constants
import { ROUTE } from '@/constants';

// Services
import { getCartByUserId } from '@/services';

export const Header = async () => {
  const userId = 134;
  const cartProduct = await getCartByUserId(userId);

  return (
    <header className="w-full md:container px-4 pt-6 md:mx-auto sm:px-6 md:px-8 lg:px-12 xl:px-16 sm:pt-8 md:pt-12 lg:pt-16 xl:pt-20">
      <div className="flex items-center justify-between gap-6">
        {/* Logo */}
        <Link href={ROUTE.PRODUCTS} className="flex items-center gap-2">
          <LogoIcon aria-label="logo" />
          <span className="text-lg font-bold w-max sm:text-xl lg:text-2xl">
            E-Comm
          </span>
        </Link>

        {/* Profile, Cart, or Menu */}
        <div className="flex items-center justify-end w-full gap-4 lg:justify-between">
          {/* Search */}
          <div className="hidden lg:block w-full xl:w-[648px]">
            <Suspense fallback={<p>...Loading</p>}>
              <SearchBox
                customClass={{
                  container: 'md:justify-end',
                  inputContainer: 'border-primary-200',
                }}
              />
            </Suspense>
          </div>

          {/* SearchBox */}
          <SearchModal />

          <div className="flex items-center gap-5">
            {/* Profile */}
            <div className="flex items-center gap-2 cursor-not-allowed">
              <ProfileIcon aria-label="profile" size={20} />
              <span className="hidden text-sm w-max md:text-base lg:text-lg lg:block">
                My account
              </span>
            </div>
            <div className="relative flex items-center">
              {/* Cart */}
              <Link href={ROUTE.CART}>
                <CartIcon
                  aria-label="cart"
                  size={30}
                  itemCount={cartProduct?.products.length ?? 0}
                />
              </Link>
            </div>
          </div>
        </div>
      </div>
    </header>
  );
};
