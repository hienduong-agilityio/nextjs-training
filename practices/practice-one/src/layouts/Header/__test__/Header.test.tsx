import { render, screen } from '@testing-library/react';

// Layouts
import { Header } from '@/layouts';

// Constants
import { ROUTE } from '@/constants';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(() => ({
    push: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
    query: {},
    asPath: '/',
  })),
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => null),
    set: jest.fn(),
  })),
}));

jest.mock('react', () => {
  const originalReact = jest.requireActual('react');
  return {
    ...originalReact,
    useOptimistic: jest.fn(() => [false, jest.fn()]),
  };
});

describe('Header Component', () => {
  it('renders the logo with a link to the products page', () => {
    const { container } = render(<Header itemCount={3} />);

    const logoLink = screen.getByRole('link', { name: /logo/i });
    expect(logoLink).toBeInTheDocument();
    expect(logoLink).toHaveAttribute('href', ROUTE.PRODUCTS);

    expect(container).toMatchSnapshot();
  });
});
