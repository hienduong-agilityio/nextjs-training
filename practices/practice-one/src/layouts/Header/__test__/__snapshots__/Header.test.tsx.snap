// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders the logo with a link to the products page 1`] = `
<div>
  <header
    class="w-full md:container px-4 pt-6 md:mx-auto sm:px-6 md:px-8 lg:px-12 xl:px-16 sm:pt-8 md:pt-12 lg:pt-16 xl:pt-20"
  >
    <div
      class="flex items-center justify-between gap-6"
    >
      <a
        class="flex items-center gap-2"
        href="/products"
      >
        <svg
          aria-label="logo"
          fill="none"
          height="44.217"
          viewBox="0 0 44 44.217"
          width="44"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="#40BFFF"
            height="44.217"
            rx="16"
            transform="scale(1)"
            width="44"
          />
          <path
            clip-rule="evenodd"
            d="M31.062 20.372c0.954 0.959 0.954 2.514 0 3.474l-7.334 7.37c-0.954 0.959-2.502 0.959-3.456 0l-7.334-7.37a2.465 2.465 0 0 1 0-3.474l7.334-7.37a2.436 2.436 0 0 1 3.457 0l7.333 7.37ZM22 18.213l-3.876 3.896L22 26.004l3.876-3.895L22 18.213Z"
            fill="#fff"
            fill-rule="evenodd"
          />
        </svg>
        <span
          class="text-lg font-bold w-max sm:text-xl lg:text-2xl"
        >
          E-Comm
        </span>
      </a>
      <div
        class="flex items-center justify-end w-full gap-4 lg:justify-between"
      >
        <div
          class="hidden lg:block w-full xl:w-[648px]"
        >
          <form
            aria-label="Input Group"
            class="flex md:justify-end"
          >
            <div
              class="w-full"
            >
              <div
                class="flex items-center gap-1 px-3 border rounded-md focus-within:ring-2 focus-within:ring-primary-100 lg:rounded-r-none border-primary-200"
              >
                <span>
                  <svg
                    aria-label="Search"
                    class="lg:hidden block"
                    fill="none"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8.875 16.75C13.2242 16.75 16.75 13.2242 16.75 8.875C16.75 4.52576 13.2242 1 8.875 1C4.52576 1 1 4.52576 1 8.875C1 13.2242 4.52576 16.75 8.875 16.75Z"
                      stroke="#40BFFF"
                      stroke-miterlimit="10"
                      stroke-width="2"
                    />
                    <path
                      d="M14.5 14.5L19 19"
                      stroke="#40BFFF"
                      stroke-linecap="round"
                      stroke-miterlimit="10"
                      stroke-width="2"
                    />
                  </svg>
                </span>
                <input
                  class="flex-1 outline-none px-3 py-2 bg-transparent disabled:cursor-not-allowed [&::-webkit-search-cancel-button]:cursor-pointer h-14 text-gray-700 appearance-none"
                  name="searchBox"
                  placeholder="Enter your query..."
                  type="search"
                  value=""
                />
              </div>
            </div>
            <button
              aria-busy="false"
              class="items-center justify-center gap-2 rounded p-2 disabled:cursor-not-allowed disabled:opacity-50 bg-opacity-100 bg-primary-200 text-white disabled:hover:bg-primary-200 hover:bg-primary-400 px-7 rounded-l-none hidden lg:flex font-semibold shadow-none"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>
        <div>
          <button
            class="block lg:hidden cursor-pointer"
            type="button"
          >
            <svg
              aria-label="Search"
              fill="none"
              height="24"
              viewBox="0 0 20 20"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.875 16.75C13.2242 16.75 16.75 13.2242 16.75 8.875C16.75 4.52576 13.2242 1 8.875 1C4.52576 1 1 4.52576 1 8.875C1 13.2242 4.52576 16.75 8.875 16.75Z"
                stroke="#9098B1"
                stroke-miterlimit="10"
                stroke-width="2"
              />
              <path
                d="M14.5 14.5L19 19"
                stroke="#9098B1"
                stroke-linecap="round"
                stroke-miterlimit="10"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <div
          class="flex items-center gap-5"
        >
          <div
            class="flex items-center gap-2 cursor-not-allowed"
          >
            <svg
              aria-label="profile"
              fill="none"
              height="20"
              viewBox="0 0 20 20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M19.356 16.2615A9.856 9.856 0 0 0 13.949 10.64a5.948 5.948 0 0 0 2.1715 -4.546a6.172 6.172 0 0 0 -1.6515 -4.105A6.1625 6.1625 0 0 0 10.495 0.0465c-0.4975 -0.062 -1.0005 -0.062 -1.498 0a6.1625 6.1625 0 0 0 -3.97 1.9435a6.172 6.172 0 0 0 -1.65 4.1035a5.949 5.949 0 0 0 2.1715 4.546A9.8455 9.8455 0 0 0 0.14 16.2615a2.354 2.354 0 0 0 0.33 2.2135A2.5885 2.5885 0 0 0 2.5505 19.5h14.405a2.5825 2.5825 0 0 0 2.078 -1.025a2.345 2.345 0 0 0 0.3225 -2.2135Zm-13.43 -10.168a3.7115 3.7115 0 0 1 1.0125 -2.48a3.7055 3.7055 0 0 1 2.417 -1.1515a3.215 3.215 0 0 1 0.79 0a3.663 3.663 0 0 1 2.9675 2.6145a3.67 3.67 0 0 1 -1.133 3.7905a3.663 3.663 0 0 1 -6.0535 -2.7735ZM2.5485 17.0625a7.766 7.766 0 0 1 2.855 -3.5415a7.7565 7.7565 0 0 1 11.55 3.5415H2.5485Z"
                fill="#22262A"
                fill-rule="evenodd"
              />
            </svg>
            <span
              class="hidden text-sm w-max md:text-base lg:text-lg lg:block"
            >
              My account
            </span>
          </div>
          <div
            class="relative flex items-center"
          >
            <a
              href="/cart"
            >
              <svg
                aria-label="cart"
                fill="none"
                height="30"
                viewBox="0 0 40 32"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 28c.447 0 .81-.354.81-.79a.801.801 0 0 0-.81-.792.801.801 0 0 0-.81.791c0 .438.362.792.81.792ZM24 28c.447 0 .81-.354.81-.79a.801.801 0 0 0-.81-.792.801.801 0 0 0-.81.791c0 .438.362.792.81.792Z"
                  fill="#22262A"
                  stroke="#22262A"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-miterlimit="10"
                  stroke-width="2.5"
                />
                <path
                  d="M4 4h3l3 18h16l3-14H10"
                  stroke="#22262A"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2.5"
                />
                <g
                  transform="translate(20 -5)"
                >
                  <circle
                    cx="8"
                    cy="10"
                    fill="#FB7181"
                    r="10"
                    stroke="white"
                    stroke-width="2"
                  />
                  <text
                    dominant-baseline="middle"
                    fill="white"
                    font-size="10"
                    font-weight="bold"
                    text-anchor="middle"
                    x="8"
                    y="10"
                  >
                    3
                  </text>
                </g>
              </svg>
            </a>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
